lib/srv/db/redis/protocol.go

```diff
@@ -0,0 +1,257 @@
+
+// subscribeCmd handles subscription in Redis. A new connection is created and maintained to read published messages
+// from Redis and send them back to the client.
+func (e *Engine) subscribeCmd(ctx context.Context, subscribeFn redisSubscribeFn, cmd *redis.Cmd) error {
+	if len(cmd.Args()) < 2 {
+		return redis.RedisError("invalid command")
+	}
+
+	args := make([]string, 0, len(cmd.Args()))
+	for _, a := range cmd.Args()[1:] {
+		args = append(args, a.(string))
+	}
+	pubSub := subscribeFn(ctx, args...)
+
+	return e.processPubSub(ctx, pubSub)
+}
+
+// processPubSub reads messages from Redis and forward them to connected client.
+func (e *Engine) processPubSub(ctx context.Context, pubSub *redis.PubSub) error {
+	for {
+		msg, err := pubSub.Receive(ctx)
+		if err != nil {
+			return err
+		}
+
+		switch msg := msg.(type) {
+		case *redis.Subscription:
+			if err := e.sendToClient([]interface{}{msg.Kind, msg.Channel, msg.Count}); err != nil {
+				return err
+			}
+		case *redis.Pong:
+			if err := e.sendToClient([]interface{}{msg.Payload}); err != nil {
+				return err
+			}
+		case *redis.Message:
+			var payloadResp interface{}
+			if len(msg.PayloadSlice) > 0 {
+				payloadResp = []interface{}{"message", msg.Channel, msg.PayloadSlice}
+			} else {
+				payloadResp = []interface{}{"message", msg.Channel, msg.Payload}
+			}
+
+			if err := e.sendToClient(payloadResp); err != nil {
+				return err
+			}
+		default:
+			err := fmt.Errorf("redis: unknown message: %T", msg)
+			return err
+		}
+	}
+}
+
+// processAuth runs RBAC check on Redis AUTH command if command contains username. Command containing only password
+// is passed to Redis. Commands with incorrect number of arguments are rejected and an error is returned.
+func (e *Engine) processAuth(ctx context.Context, redisClient redis.UniversalClient, cmd *redis.Cmd) error {
+	// AUTH command may contain only password or login and password. Depends on the version we need to make sure
+	// that the user has permission to connect as the provided db user.
+	// ref: https://redis.io/commands/auth
+	switch len(cmd.Args()) {
+	case 1:
+		// Passed AUTH command without any arguments. Mimic Redis error response.
+		return errors.New("wrong number of arguments for 'auth' command")
+	case 2:
+		// Old Redis command. Password is the only argument here. Pass to Redis to validate.
+		// ex. AUTH my-secret-password
+		return redisClient.Process(ctx, cmd)
+	case 3:
+		// Redis 6 version that contains username and password. Check the username against our RBAC before sending to Redis.
+		// ex. AUTH bob my-secret-password
+		dbUser, ok := cmd.Args()[1].(string)
+		if !ok {
+			return errors.New("username has a wrong type, expected string")
+		}
+
+		err := e.sessionCtx.Checker.CheckAccess(e.sessionCtx.Database,
+			services.AccessMFAParams{Verified: true},
+			role.DatabaseRoleMatchers(
+				defaults.ProtocolRedis,
+				dbUser,
```

Comment:
This will work for most of the cases though there are some corner scenarios like MFA access where db check is done based on user name encode in identity 
https://github.com/gravitational/teleport/blob/jakule/redis-support/lib/auth/auth.go#L3264

Reply:
Look at the latest changes. I think this is covered now.